### DISCUSSION TITLE
Improvements for search

### DISCUSSION THREAD

**Discussion by k4nzdroid** [zim-desktop-wiki/zim-desktop-wiki#2398] [Ideas]
**Allow users to choose the default search keyword.**

In our scenarios, we know which page we want. It is enough to find the page by searching its name or location. There is no need to search throught the whole page contents, which would slow down the search speed. 

Currently, `contentorname` is the default search keyword. To find a page quickly, we have to type the `Name:` serval times. For example, if we need to find a page named 'SSH Login with Password', we have to type `Name: ssh Name: login Name: passowrd`. Just inputting `ssh login name` would be better.

When we deal with unfamiliar notebooks, we may need to search page content to find pages.

So, It would be better to allow users to choose the default search keyword.

**Implementing search as a plugin.**

We tried to add an option for searching, but the search function is not a plugin.  

There are to two ways to do that:
* Adding the option to configuration files. While it is easy to implement, it is hard to use.
* Adding the option to Preference. However, it is only related to search, which makes the preferences cluttered.

While search is an important feature, it may no be essential enough to be considered a core funtion. It is more appropriate to have is as a separate plugin. Additionally, making its configuration independent would allow for greater flexibility and customization. 

**Considering for command line searching.**

Add a way to allow plugins to run in command line. 

> Maybe Zim should be a framework - an abstraction of note elements, such as Page, Tag, Link, and so on. And it should be surrounded by plugins, which manipulate those elements :)

**Comment by jaap-karssenberg** [comment]
I think for this use case maybe you should not need to use search at all,
but just an improved version of the completion in the "jump to" dialog.

Making search plugin extendable is definitely on my whish list.

Plugins can define commandline commands - search for "Command" classes in
the plugin folder for examples :)

**Comment by k4nzdroid** [comment]
aha..."jump to" is a good choice, we will try to improve the completion. 🥇

**Comment by k4nzdroid** [comment]
https://github.com/zim-desktop-wiki/zim-desktop-wiki/pull/2400
